lib/SILOptimizer/Transforms/CopyForwarding.cpp

```diff
@@ -1166,6 +1167,118 @@ void CopyForwarding::forwardCopiesOf(SILValue Def, SILFunction *F) {
 }
 
 //===----------------------------------------------------------------------===//
+//                    Temporary RValue Optimization
+//===----------------------------------------------------------------------===//
+
+struct TempRValue {
+  CopyAddrInst *copy = nullptr;
+  DestroyAddrInst *destroy = nullptr;
+  DeallocStackInst *dealloc = nullptr;
+};
+
+/// Peephole optimization to eliminate short-lived immutable temporary copies.
+/// This handles a common pattern generated by SILGen where temporary RValues
+/// are emitted as copies...
+///
+/// @foo : (@inout NonLoadable) -> ()
+/// bb(%0 : $*NonLoadable):
+///   %2 = struct_element_addr %0 : $*S<B>, #S._value
+///   %3 = load %2 : $*Builtin.Int64
+///   %4 = alloc_stack $S<B>
+///   copy_addr %1 to [initialization] %4 : $*S<B>
+///   %6 = struct_element_addr %4 : $*S<B>, #S._value
+///   %7 = load %6 : $*Builtin.Int64
+///   %8 = builtin "cmp_slt_Int64"(%3 : $Builtin.Int64, %7 : $Builtin.Int64)
+///        : $Builtin.Int1
+///   destroy_addr %4 : $*S<B>
+///   dealloc_stack %4 : $*S<B>
+///
+/// This differs from the copy forwarding algorithm below because it handles
+/// copy source and dest lifetimes that are unavoidably overlappying. Instead,
+/// it finds cases in which it is easy to determine that the source is
+/// unmodified during the copy destination's lifetime. Thus, the destination can
+/// be viewed as a short-lived "rvalue".
+///
+/// The optimization is identified in two parts.
+///
+/// 1. The copy's destination is an AllocStack in which all all users are of the
+/// copied value. i.e. They all access the value initialized at the copy and
+/// destroyed by the subsequent destroy. This is sufficient to prove that the
+/// copy can be safely removed via replace-all-uses. i.e. there are no
+/// undetected aliases into the copy's destination memory.
+///
+/// 2. No instructions between the copy and the destroy of its destination
+/// writes to memory. This is sufficient to prove the copy is unnecesary.
+///
+/// Return the single dealloc_stack corresponding to the temporary storage for
+/// teh copy's destination. Return null if the copy does not initialize an
+/// optimizable temporary.
+static Optional<TempRValue> canOptimizeTempRValue(CopyAddrInst *copyInst) {
+  if (copyInst->isTakeOfSrc() || !copyInst->isInitializationOfDest())
+    return None;
+
+  auto *copyDest = dyn_cast<AllocStackInst>(copyInst->getDest());
+  if (!copyDest)
+    return None;
+
+  // Scan all users to verify they refer to the copy's destination value. It is
+  // sufficient to check that the only users that modify memory are the
+  // copy_addr [initialization] and destroy_addr.
+  TempRValue rvalue;
+  rvalue.copy = copyInst;
+  for (auto *useOper : copyDest->getUses()) {
```

Comment:
Checking if the user is in the same block as the copy_addr here might save some compile time.
Consider a very large basic block with lots of copy_addrs where this is not the case. You would iterate over all instructions of the basic block for all copy_addrs (which is quadratic).